File: frontend/js/event.js
// Add event listener for the event form submission
document.getElementById('eventForm')?.addEventListener('submit', function (e) {
    e.preventDefault();

    const eventId = document.getElementById('event-id').value;
    const event = {
        title: document.getElementById('title').value,
        description: document.getElementById('description').value,
        date: document.getElementById('date').value,
        time: document.getElementById('time').value,
        location: document.getElementById('location').value,
        price: document.getElementById('price').value
    };

    const url = eventId
        ? `http://localhost:5000/events/${eventId}`
        : 'http://localhost:5000/create';

    const method = eventId ? 'PUT' : 'POST';

    fetch(url, {
        method: method,
        headers: {
            'Content-Type': 'application/json'
        },
        body: JSON.stringify(event)
    })
        .then(response => response.json())
        .then(data => {
            alert(eventId ? 'Event updated successfully!' : 'Event created successfully!');
            window.location.href = 'dashboard.html'; // Redirect to dashboard after successful operation
        })
        .catch(error => console.error('Error:', error));
});

// Global variables to hold fetched events and purchased events
let events = [];
let purchasedEvents = [];

// Load events from the server and then fetch purchased tickets
function loadEvents() {
    const userId = localStorage.getItem('userId');
    if (!userId) {
        console.error('No user ID found in local storage. User might not be logged in.');
        return;
    }

    Promise.all([
        fetch('http://localhost:5000/events').then(response => response.json()),
        fetch(`http://127.0.0.1:5000/user/${userId}/tickets`).then(response => response.json())
    ])
    .then(([eventsData, purchasedEventsData]) => {
        events = eventsData;
        purchasedEvents = purchasedEventsData;
        displayEvents(events);
        displayPurchasedEvents();
    })
    .catch(error => console.error('Error loading data:', error));
}

// Display all events on the page
function displayEvents(eventsToDisplay) {
    const eventList = document.getElementById('dashboard');
    eventList.innerHTML = ''; // Clear existing content

    eventsToDisplay.forEach(event => {
        const eventCard = createEventCard(event);
        eventList.appendChild(eventCard);
    });
}

// Display purchased events
function displayPurchasedEvents() {
    const purchasedEventsList = document.getElementById('purchased-events');
    purchasedEventsList.innerHTML = ''; // Reset content

    if (purchasedEvents.length === 0) {
        purchasedEventsList.innerHTML = '<p>No purchased events yet.</p>';
    } else {
        purchasedEvents.forEach(event => {
            const eventCard = createEventCard(event, true);
            purchasedEventsList.appendChild(eventCard);
        });
    }
}

// Create an event card
function createEventCard(event, isPurchased = false) {
    const eventCard = document.createElement('div');
    eventCard.className = 'event-card';
    eventCard.setAttribute('data-event-id', event.id);

    eventCard.innerHTML = `
        
        <div class="event-content">
            <h3 class="event-title">${event.title}</h3>
            <p class="event-date">Date: ${event.date}</p>
            <p class="event-time">Time: ${event.time}</p>
            <p class="event-location">Location: ${event.location}</p>
            <p class="event-price">Price: $${event.price}</p>
            <p class="event-description">${event.description}</p>
            ${isPurchased ?
                '<p class="purchase-status">Purchased</p>' :
                `<a href="#" class="event-button" data-event-id="${event.id}">Buy Ticket</a>`
            }
        </div>
    `;

    if (!isPurchased) {
        const buyTicketButton = eventCard.querySelector('.event-button');
        buyTicketButton.addEventListener('click', (e) => {
            e.preventDefault();
            buyTicket(event.id, event.price);
        });
    }

    return eventCard;
}

// Filter events based on search input
function filterEvents() {
    const searchTerm = document.getElementById('search-bar').value.toLowerCase();
    const filteredEvents = events.filter(event => {
        return (
            event.title.toLowerCase().includes(searchTerm) ||
            event.description.toLowerCase().includes(searchTerm) ||
            event.location.toLowerCase().includes(searchTerm)
        );
    });
    displayEvents(filteredEvents);
}

// Function to buy a ticket for a specific event
function buyTicket(eventId, price) {
    const userId = localStorage.getItem('userId');

    if (!userId) {
        alert('You must be logged in to purchase a ticket.');
        return;
    }

    const ticketData = {
        event_id: eventId,
        user_id: userId,
        price: price,
        status: 'booked'
    };

    fetch('http://127.0.0.1:5000/tickets', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify(ticketData),
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            alert('Ticket purchased successfully!');
            loadEvents();
        } else {
            if (data.error === 'duplicate_ticket') {
                alert('You have already purchased a ticket for this event.');
            } else {
                alert('Failed to purchase ticket: ' + data.message);
            }
        }
    })
    .catch((error) => {
        console.error('Error:', error);
        alert('An error occurred while purchasing the ticket. Please try again.');
    });
}


document.addEventListener('DOMContentLoaded', loadEvents);


document.getElementById('search-bar').addEventListener('keyup', filterEvents);

File: backend/event.py
from flask import request, jsonify
from models import Event,Ticket  
from db import db  

def create_event():
    data = request.json  

 
    title = data.get('title')
    description = data.get('description')
    date = data.get('date')
    time = data.get('time')
    location = data.get('location')
    price = data.get('price')

   
    new_event = Event(title=title, description=description, date=date, time=time, location=location, price=price)
    db.session.add(new_event)
    db.session.commit()

    return jsonify({"message": "Event created successfully", "event_id": new_event.id}), 201


def get_all_events():
    events = Event.query.all()  
    return jsonify([{
        "id": event.id,
        "title": event.title,
        "description": event.description,
        "date": event.date.isoformat(),  
        "time": event.time.isoformat(),  
        "location": event.location,
        "price": event.price
    } for event in events]), 200



def delete_event():
    print("Received DELETE request")
    print("Request headers:", dict(request.headers))

    try:
        data = request.get_json(silent=True)
        print("Received data:", data)

        if not data:
            print("No JSON data received")
            return jsonify({"message": "No JSON data received"}), 400

        if 'event_id' not in data:
            print("No event_id in data")
            return jsonify({"message": "Missing event_id in request"}), 400

        event_id = data.get('event_id')
        print(f"Attempting to delete event with ID: {event_id}")

      
        event = Event.query.get(event_id)
        if not event:
            print(f"Event with id {event_id} not found")
            return jsonify({"message": f"Event with id {event_id} not found"}), 404

        db.session.delete(event)
        db.session.commit()
        print(f"Successfully deleted event {event_id}")
        return jsonify({"message": "Event deleted successfully"}), 200
    except Exception as e:
        print(f"Error occurred: {str(e)}")
        db.session.rollback()
        return jsonify({"message": f"Error deleting event: {str(e)}"}), 500


def edit_event(event_id):
    try:
        event = Event.query.get(event_id)
        if not event:
            return jsonify({"message": f"Event with id {event_id} not found"}), 404

        data = request.get_json()

       
        if 'title' in data:
            event.title = data['title']
        if 'description' in data:
            event.description = data['description']
        if 'date' in data:
            event.date = data['date']
        if 'time' in data:
            event.time = data['time']
        if 'location' in data:
            event.location = data['location']
        if 'price' in data:
            event.price = data['price']

        db.session.commit()

        
        return jsonify({
            "message": "Event updated successfully",
            "event": {
                "id": event.id,
                "title": event.title,
                "description": event.description,
                "date": event.date,
                "time": event.time,
                "location": event.location,
                "price": event.price
            }
        }), 200
    except Exception as e:
        db.session.rollback()
        return jsonify({"message": f"Error updating event: {str(e)}"}), 500




def create_ticket():
    data = request.json
    event_id = data.get('event_id')
    user_id = data.get('user_id')
    price = data.get('price')

  
    existing_ticket = Ticket.query.filter_by(event_id=event_id, user_id=user_id).first()
    if existing_ticket:
        return jsonify({'success': False, 'message': 'Event already purchased', 'error': 'duplicate_ticket'}), 400

    new_ticket = Ticket(event_id=event_id, user_id=user_id, price=price, status='booked')

    try:
        db.session.add(new_ticket)
        db.session.commit()
        return jsonify({'success': True, 'message': 'Ticket booked successfully!', 'ticket_id': new_ticket.id})
    except IntegrityError:
        db.session.rollback()
        return jsonify({'success': False, 'message': 'Error booking ticket', 'error': 'database_error'}), 500
    except Exception as e:
        db.session.rollback()
        return jsonify({'success': False, 'message': str(e), 'error': 'unknown_error'}), 500

def get_purchased_tickets(user_id):
    tickets = Ticket.query.filter_by(user_id=user_id).all()
    purchased_event_ids = [ticket.event_id for ticket in tickets]

    
    purchased_events = Event.query.filter(Event.id.in_(purchased_event_ids)).all()

    
    events_data = []
    for event in purchased_events:
        events_data.append({
            'id': event.id,
            'title': event.title,
            'description': event.description,
            'date': event.date.strftime('%Y-%m-%d'), 
            'time': event.time.strftime('%H:%M'),     
            'location': event.location,
            'price': float(event.price), 

        })

    return jsonify(events_data)
